PROF-153 discussion (#106)
SW805F20/report: sections/conclusion/discussion.tex

Along with the questions, the stand-up meetings were changed to be conducted daily, which further facilitated the developers staying focused.
\\\\
\todo{pivoting and essence}

\subsection{Reflecting on mobility}
\todo{VI SKAL HAVE INTRODUCERET NOGET MOBILITY. NÅR DET ER GJORT SKAL DER VÆRE NOGET REF TIL DET HER.}
This system achieves the goal of mobility in two separate ways.
The main aspect of the system as a whole is the game itself.
This game is developed for mobile devices, and achieves mobility in the sense that the application can be used by clients in many different locations, rather than being restricted to a certain location.
One issue with this, however, is the Python based host application.
This application is currently developed for use on a computer, meaning it is not as mobile as the game counterpart.
It is not fully restricted to a certain location, however, as it can be executed on a laptop which can be moved to the location of the playing field.
\\\\
The game requires an internet connection to be played.

> This could also limit mobility, but during the test \todo{ref til st afsnit 18 maj} the game was played outdoors using a mobile network.

Reviewer: Den er på develop nu, så du kan godt lave en autoref til den

The communication should have been more clear, and more care should have been put into continuous testing of the system.
These issues led to the necessity of conducting a second test, described in \todo{autoref til 18 maj}, for which more care was put into preparation to ensure the generation of results.

\subsection{Comparing the final product to the MVP}
In \autoref{sec:mvp} an MVP was defined for a viable game solution, split into functional and non-functional requirements.

\subsubsection{Functional requirements}
The functional requirements defined the core features of the game.
The first three requirements are defined as starting the game, connecting clients and server and users viewing their position.
All of this is accomplished through the network protocol defined in \autoref{app:network}, specifically TCP message 3 for game start and UDP message 0 to update player positions.
Connecting players is done through the lobby screen based on IP.
The next few requirements deal with actually playing the game, requiring the player to be able to view a playing field, goals, the players, the ball and current score, as well as being able to win a game based on their current score.

> This has all been accomplished in the Unity game, as shown in REFREFREFREF.

Reviewer 2: ????

\todo{tænker vi kan ref til sprint-5 coinclusion til at vise det. Der skal vel være noget afsluttende ift spillet}
Finally, the game should be dynamic, allowing for an arbitrary number of players and goals to win as well as different sizes of playing fields.
This is partly accomplished.
The game supports an arbitrary number of goals to win, however, in terms of players there is currently a maximum capacity of four players in a single game.
This is communicated by the host to the clients through TCP message 6 in \autoref{app:network}.
The playing field scales based on the user input, meaning that requirement is supported, but fields have to be shaped like a rectangle or a square in order to support the current implementation of generating new goal zones.

\subsubsection{Non-functional requirements}

> The non-functional requirements to be evaluated through a usability test.

Reviewer: ```suggestion
The non-functional requirements were ideally to be evaluated through a usability test.
However, due to the circumstances, this was not possible.
```

> A usability test was conducted in \todo{ref til 18 maj}, but it could not be conducted properly due to the circumstances.

Reviewer: ```suggestion
Instead a more unstructured test was conducted with the group as described in \todo{ref til 18 maj}.
```

However, the non-functional requirements state that the game should be pleasant to look at and not cause discomfort for players, and there should be enough updates to represent the real positions of the players.
None of the testers experienced nausea or discomfort during the playing of the game, indicating that the first requirement was likely achieved.
There were only six testers however, which is likely not enough testers to determine this completely.
The testers could have been outliers, or there could be possible outliers who would find the playing of the game uncomfortable.

> As explained in \todo{18 maj ref}, there were some slight issues in terms of updates when playing the game outside.

Reviewer: tilføj autoref

Occasionally, updates would stop being sent for a period of several seconds.
This caused play to be interrupted.

> \todo{Den her sætning herunder. Er det godt nok sådan her? Eller er det at vi blamer tags?}

Reviewer 2: Altså vi der kunne måske godt spekuleres i forskellige årsager til det her men nej vi ved jo ikke hvad der er skyld i det

Reviewer 2: Der kan evt. tilføjes noget med at vi har forsøgt at gøre op for den lave update rate ved implementering af lerping når det engang er merged ind

> The reason for this issue has not been pinpointed, and it could have been caused by an unstable connection to the mobile network.

Reviewer: Vi modtog jo ikke opdateringerne på netværket til hosten, men vi konkluderer det ikke rigtigt. Det er nok meget fint.

Development was segregated into five sprints, defining the different iterations of the system.
These sprints served as a guideline for when to have sprint retrospectives and enact changes in the process.
Each retrospective would cause slight alterations to the process.
\\\\
Initially, tasks were estimated based on planning poker.
However, we found it difficult to properly estimate a lot of the tasks, as this project concerned itself with an area of expertise we were not familiar with.
This led to the inclusion of an Essence based version of estimating in sprint 1.
This version of estimation was also scrapped, as described in \autoref{sec:sprint2-conclusion}, due to it seeming biased towards shorter tasks and estimation not being accurate.
This led to the method of simply having a group-based discussion about the DoD of a task, and ensuring everyone understood it.
This DoD based task assignment worked well for the final sprints.
It avoided spending unnecessary time on planning poker on top of discussing tasks, and made the developers more productive.
\\\\
Another aspect improved by the retrospectives was the daily stand-up meetings.

> In the beginning of the project these tended to be unfocused and not provide the necessary information.

Reviewer: Smid evt også en sætning ind omkring at daily stand up ikke var daily i starten. Men at det senere i processen blev det.

One issue with this, however, is the Python based host application.
This application is currently developed for use on a computer, meaning it is not as mobile as the game counterpart.
It is not fully restricted to a certain location, however, as it can be executed on a laptop which can be moved to the location of the playing field.
\\\\
The game requires an internet connection to be played.
This could also limit mobility, but during the test \todo{ref til st afsnit 18 maj} the game was played outdoors using a mobile network.
Doing so facilitates both indoor and outdoor play, further improving the mobility of the system in the sense that it can be used in different conditions.
Another way this system achieves the mobility aspect is that, when in use, the game requires players to be mobile.
While players are mobile when using the system, the Pozyx system is used to provide a service in regards to location information which is communicated wirelessly by the underlying network protocol.

> 

Reviewer: ```suggestion
```

Reviewer: Fuck af med dit linjefarm

Author: fuck af med dit liv

\subsection{The final design}
no lobby, bad
gonna add stuff here after code stop i guess

\subsection{UDP and TCP for the protocol}

> when the socket type to be used for the program was initially chosen in \autoref{sec:sprint1-networking}, a large emphasis was put on the rate at which we could send messages.

Reviewer: ```suggestion
When the socket type to be used for the program was initially chosen in \autoref{sec:sprint1-networking}, a large emphasis was put on the rate at which we could send messages.
```

The sending rate also affected a lot of the other decisions made for the networking protocol, such as keeping messages as short as possible.
The goal was to facilitate sending as many messages as possible.
However, once testing began, it became obvious that this might not have been the fully correct approach.
While running the host with mock data, it was possible to more or less crash our connections.
\\\\
In order to fix this, we implemented some message throttling, described in \autoref{sec:sprint5-transmissionlimit}, based on the maximum amount of updates from the available Pozyx system.
Keeping this limiting of messages in mind, the aspect of sending messages rapidly loses some of its importance, leading to some of our earlier considerations not being optimal.
This might have led us to implement a TCP aspect earlier in the project, avoiding having to spend some of the time attempting to implement UDP acknowledgments. 

\subsection{Testing the system}

> The first full scale test of the system and all its components took play on May 7th, as described in \autoref{sec:initial-test}.

Reviewer: ```suggestion
The first full scale test of the system and all its components took place on May 7th, as described in \autoref{sec:initial-test}.
```

All of this is accomplished through the network protocol defined in \autoref{app:network}, specifically TCP message 3 for game start and UDP message 0 to update player positions.
Connecting players is done through the lobby screen based on IP.
The next few requirements deal with actually playing the game, requiring the player to be able to view a playing field, goals, the players, the ball and current score, as well as being able to win a game based on their current score.
This has all been accomplished in the Unity game, as shown in REFREFREFREF.
\todo{tænker vi kan ref til sprint-5 coinclusion til at vise det. Der skal vel være noget afsluttende ift spillet}
Finally, the game should be dynamic, allowing for an arbitrary number of players and goals to win as well as different sizes of playing fields.
This is partly accomplished.
The game supports an arbitrary number of goals to win, however, in terms of players there is currently a maximum capacity of four players in a single game.
This is communicated by the host to the clients through TCP message 6 in \autoref{app:network}.

> The playing field scales based on the user input, meaning that requirement is supported, but fields have to be shaped like a rectangle or a square in order to support the current implementation of generating new goalzones.

Reviewer: ```suggestion
The playing field scales based on the user input, meaning that requirement is supported, but fields have to be shaped like a rectangle or a square in order to support the current implementation of generating new goal zones.
```
